### PR TITLE
fix(gatsby-plugin-offline): delay adding resources for paths until we have urls

### PR DESCRIPTION
**Pull request by davidbailey00**
This adds a new API called `onPostPrefetchPathname`, which runs *after* resources have been fetched for a pathname, as opposed to `onPrefetchPathname` which runs directly upon fetching resources.

In plugin-offline, the existing API `onPrefetchPathname` was being used - this was causing problems, since the resources weren't always fetched in time for `getResourceURLsForPathname` to return the data (see https://github.com/gatsbyjs/gatsby/pull/8613/files#diff-450f89b4ad11867864b3d8d21f83e442R35 - getResourceURLsForPathname sometimes returned null).

Now that the API `onPostPrefetchPathname` is used instead, `getResourceURLsForPathname` will always work properly, since the API runs after the resources are fetched and only if they are fetched successfully.

In addition to this, I've passed `onPostPrefetchPathname` into the API runner for `onPrefetchPathname` so that plugins which use custom prefetching can run this API at the right time.

### CODE REVIEW

**Comment by davidbailey00** [comment]
AppVeyor why...

image

**Comment by KyleAMathews** [comment]
This is called when the "prefetching" has _started_ not when the resources has actually been downloaded.

Can you explain more about the error plugin-offline is causing?

**Comment by davidbailey00** [comment]
@KyleAMathews We need to guarantee that `getResourceURLsForPathname` will actually work e.g. in gatsby-plugin-offline (see updated description) - currently it's sometimes returning null, since the URLs have not been populated in time, or it returns null if the pathname is a 404 (e.g. integration-tests/production-runtime for "/page-3").

With the new API, it's called at a time when `getResourceURLsForPathname` is guaranteed to work, so there's no longer an error like "Cannot call forEach of null` in plugin-offline (again at the link in the updated description).

File: packages/gatsby/cache-dir/loader.js [hunk 1]
```diff
@@ -236,6 +236,9 @@ const queue = {
       prefetchResource(page.componentChunkName)
     }
 
+    // Tell plugins the path has been successfully prefetched
+    apiRunner(`onPostPrefetchPathname`, { pathname: path })
```
Comment (pieh) [inline packages/gatsby/cache-dir/loader.js]: it this point we don't know if resources are prefetched yet - only that we started prefetching already I think?

File: packages/gatsby-plugin-guess-js/src/gatsby-browser.js [hunk 2]
```diff
@@ -69,6 +73,8 @@ exports.onPrefetchPathname = ({ pathPrefix }, pluginOptions) => {
           resources.push(`static/d/${___dataPaths[page.jsonName]}.json`)
           // TODO add support for pathPrefix
           resources.forEach(r => prefetch(`/${r}`))
+
+          onPostPrefetchPathname({ pathname: p })
```
Comment (pieh) [inline packages/gatsby-plugin-guess-js/src/gatsby-browser.js]: Could we refactor this a bit so `loader.js` exposes `prefetchResourcesForPath` function and `guess` plugin and default gatsby prefetching use same implementation (and `onPostPrefetchPathname` api is called just from `loader.js`) ?

**Pull request merged**